add notes for n27145 (#685)
bitcoin-core-review-club/website: _posts/2023-05-31-#27145.md

---
layout: pr
date: 2023-05-31
title: "When a block is disconnected, update transactions that are no longer conflicted"
pr: 27145
authors: [ishaanam]
components: ["wallet"]
host: josibake
status: upcoming

> commit:

Reviewer: ```suggestion
commit: 89df798
```

  used to calculate the wallet's balance(s). For example:

    - If a transaction that is 100 blocks deep in the most-work chain, the wallet can
      reasonably include its UTXOs in the balance displayed to the user.

    - If a transaction conflicts with another transaction 100 blocks deep in the most-work chain,
      the wallet can be equally sure that, even though the transaction may have a valid signature,
      its UTXOs do not count towards the user's balance.

    - If a transaction is unconfirmed and in the node's mempool, the wallet should account for its
      UTXOs, but not consider them as safe as confirmed ones.

- The author has provided more notes on transaction states and their effects on balance calculation [here](https://gist.github.com/ishaanam/846adf3b453c3a85fe6e15c882c57ae0#locations-where-tx-states-are-updated).

- [Wallet Transaction Conflict
  Tracking](https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking)
across chainstate and mempool events is tricky. As described in [Issue #7315](https://github.com/bitcoin/bitcoin/issues/7315),

> when a block is disconnected, the wallet should be marking conflicted transactions as inactive but
> isn't.  [PR #27145](https://github.com/bitcoin/bitcoin/pull/27145) updates the behavior to mark

Reviewer: nit
```suggestion
when a block is disconnected, the wallet should be marking conflicted transactions as inactive, but
isn't currently doing so. [PR #27145](https://github.com/bitcoin/bitcoin/pull/27145) updates the behavior to mark
```

> isn't.  [PR #27145](https://github.com/bitcoin/bitcoin/pull/27145) updates the behavior to mark
> conflicted transactions after a reorg inactive.

Reviewer: ```suggestion
isn't.  [PR #27145](https://github.com/bitcoin/bitcoin/pull/27145) updates the behavior to mark
transactions that are no longer conflicting after a reorg as inactive.
```

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?

> 1. What is the issue this PR addresses? Can you reproduce the problem described? (Hint: try running

Reviewer 2: Hint: try running  wallet\_conflicts.py on master. What error do you get?)

This PR has been merged and testing on master is the same behaviour

Reviewer: `master` is relative indeed. In the review club, generally that refers to `HEAD` on `master` prior to the PR being merged. Since the notes usually reference the [PR `HEAD` that we're considering](https://github.com/bitcoin-core-review-club/website/blob/3a306ef674faa39fe36103f1e6af6a1d41f903cb/_posts/2023-05-31-%2327145.md?plain=1#L10), you can retroactively find the `master` branch at the time of writing by looking at the `git log`, e.g. in this case it's `49d07ea9a`.

```
git checkout 89df798
git log --oneline
```

Note: we keep track of the PR `HEAD`s on the `bitcoin-core-review-club/bitcoin` repo to make sure the commits don't get purged, so you can also find them at https://github.com/bitcoin-core-review-club/bitcoin/tags

Does that resolve your issue?

   wallet\_conflicts.py on master. What error do you get?)

1. What are the different
   [states](https://github.com/bitcoin/bitcoin/blob/25202cace9140870c75cb3a811e10045df88c226/src/wallet/transaction.h#L243)
a `CWalletTx` can be in?

1. Which, if any, of the `TxState`s are "final," i.e. once a transaction reaches this
state, it will never change again?

> 1. Where in net\_processing.cpp triggers `CWallet::blockDisconnected()` to be executed (Hint: it is

Reviewer: grammar nit
```suggestion
1. Where in net\_processing.cpp is `CWallet::blockDisconnected()` triggered to be executed (Hint: it is
```
